diff -ruN a/src/alloc/alloc.h src/alloc/alloc.h
--- a/src/alloc/alloc.h	2026-10-05 11:31:06.650999049 +0200
+++ b/src/alloc/alloc.h	2026-10-05 12:07:22.270482681 +0200
@@ -67,7 +67,6 @@
 // Checks if the page pointed at by `ptr` is huge. Assumes that `ptr` has already
 // been allocated.
 static void check_huge_page(void *ptr) {
-	const uint64_t CUSTOM_PAGE_SIZE = 1u<<13; // TODO dont know if this is correct
 	int pagemap_fd = open("/proc/self/pagemap", O_RDONLY);
 	if (pagemap_fd < 0) {
 		std::cout << "could not open /proc/self/pagemap: " << strerror(errno) << "\n";
diff -ruN a/src/atomic/atomic_primitives.h src/atomic/atomic_primitives.h
--- a/src/atomic/atomic_primitives.h	2026-10-05 11:31:06.651751839 +0200
+++ b/src/atomic/atomic_primitives.h	2026-10-05 12:04:17.533064989 +0200
@@ -244,6 +244,9 @@
 	__ATTRIBUTE__(no_thread_safety_analysis)
 
 
+#ifndef _GLIBCXX_ALWAYS_INLINE
+#define _GLIBCXX_ALWAYS_INLINE inline __attribute__((__always_inline__))
+#endif
 namespace __atomic_impl {
 
 	// Remove volatile and create a non-deduced context for value arguments.
@@ -252,7 +255,7 @@
 
 	// Like _Val<T> above, but for difference_type arguments.
 	template<typename _Tp>
-	using _Diff = std::__conditional_t<std::is_pointer_v<_Tp>, ptrdiff_t, _Val<_Tp>>;
+	using _Diff = std::conditional_t<std::is_pointer_v<_Tp>, ptrdiff_t, _Val<_Tp>>;
 
 	// Implementation details of atomic padding handling
 	/// \tparam T
@@ -270,7 +273,7 @@
 	/// \return TODO
 	template<typename _Tp>
 	inline _Tp *__clear_padding(_Tp &__val) noexcept {
-		auto *__ptr = std::__addressof(__val);
+		auto *__ptr = __builtin_addressof(__val);
 		if constexpr (__atomic_impl::__maybe_has_padding<_Tp>()) {
 			__builtin_clear_padding(__ptr);
 		}
@@ -294,7 +297,7 @@
 			alignas(_Vp) unsigned char __buf[sizeof(_Vp)];
 			_Vp *__exp = ::new ((void *) __buf) _Vp(__e);
 			__atomic_impl::__clear_padding(*__exp);
-			if (__atomic_compare_exchange(std::__addressof(__val), __exp,
+			if (__atomic_compare_exchange(__builtin_addressof(__val), __exp,
 			                              __atomic_impl::__clear_padding(__i),
 			                              __is_weak, int(__s), int(__f))) {
 				return true;
@@ -363,21 +366,37 @@
 	_GLIBCXX_ALWAYS_INLINE void
 	wait(const _Tp *__ptr, _Val<_Tp> __old,
 	     std::memory_order __m = std::memory_order_seq_cst) noexcept {
+#ifdef __GLIBCXX__
 		std::__atomic_wait_address_v(__ptr, __old,
 		                             [__ptr, __m]() { return __atomic_impl::load(__ptr, __m); });
+#else
+		std::atomic_ref<_Tp>(*const_cast<_Tp *>(__ptr)).wait(__old, __m);
+#endif
 	}
 
 	// TODO add const volatile overload
 
 	template<typename _Tp>
 	_GLIBCXX_ALWAYS_INLINE void
-	notify_one(const _Tp *__ptr) noexcept { std::__atomic_notify_address(__ptr, false); }
+	notify_one(const _Tp *__ptr) noexcept {
+#ifdef __GLIBCXX__
+		std::__atomic_notify_address(__ptr, false);
+#else
+		std::atomic_ref<_Tp>(*const_cast<_Tp *>(__ptr)).notify_one();
+#endif
+	}
 
 	// TODO add const volatile overload
 
 	template<typename _Tp>
 	_GLIBCXX_ALWAYS_INLINE void
-	notify_all(const _Tp *__ptr) noexcept { std::__atomic_notify_address(__ptr, true); }
+	notify_all(const _Tp *__ptr) noexcept {
+#ifdef __GLIBCXX__
+		std::__atomic_notify_address(__ptr, true);
+#else
+		std::atomic_ref<_Tp>(*const_cast<_Tp *>(__ptr)).notify_all();
+#endif
+	}
 
 	// TODO add const volatile overload
 #endif// __cpp_lib_atomic_wait
diff -ruN a/src/helper.h src/helper.h
--- a/src/helper.h	2026-10-05 11:31:06.666925946 +0200
+++ b/src/helper.h	2026-10-05 12:07:22.270211599 +0200
@@ -1,6 +1,19 @@
 #ifndef CRYPTANALYSISLIB_HELPER_H
 #define CRYPTANALYSISLIB_HELPER_H
 
+// NOTE: needed to rename `PAGE_SIZE` to `CUSTOM_PAGE_SIZE`, as apple in
+// its infinite wisdom have a global variable called `PAGE_SIZE`.
+#ifdef FORCE_HPAGE
+// normal page, 4KiB, buts its forced to be an huge page
+#define CUSTOM_PAGE_SIZE (1 << 21)
+#else
+// normal page, 4KiB
+#define CUSTOM_PAGE_SIZE (1 << 12)
+#endif
+
+// huge page, 2MiB
+#define HPAGE_SIZE (1 << 21)
+
 // Global Includes
 #include <cassert>
 #include <cstdint>// needed for uint8_t and so on
diff -ruN a/src/simd/neon.h src/simd/neon.h
--- a/src/simd/neon.h	2026-10-05 11:31:06.680722222 +0200
+++ b/src/simd/neon.h	2026-10-05 12:04:26.773373230 +0200
@@ -3275,40 +3275,6 @@
 		return ret;
 	}
 
-	constexpr static inline int lt(const S in1,
-                                   const S in2) noexcept {
-		uint32_t ret = 0;
-
-		LOOP_UNROLL()
-		for (uint32_t i = 0; i < 2; ++i) {
-#ifdef __clang__
-			const uint32x4_t tmp = vcltq_u32(in1.v128[i], in2.v128[i]);
-			ret ^= _mm_movemask_epi32(tmp) << i * 4;
-#else
-			const uint32x4_t tmp = in1.v128[i] < in2.v128[i];
-			ret ^= _mm_movemask_epi32(tmp) << i * 4;
-#endif
-		}
-
-		return ret;
-	}
-
-	constexpr static inline S lt_(const S in1,
-                                  const S in2) noexcept {
-		S ret;
-
-		LOOP_UNROLL()
-		for (uint32_t i = 0; i < 2; ++i) {
-#ifdef __clang__
-			ret.v128[i] = vcltq_u32(in1.v128[i], in2.v128[i]);
-#else
-			ret.v128[i] = in1.v128[i] < in2.v128[i];
-#endif
-		}
-
-		return ret;
-	}
-
 	constexpr static inline int le(const S in1,
                                    const S in2) noexcept {
 		uint32_t ret = 0;
diff -ruN a/src/thread/performance.h src/thread/performance.h
--- a/src/thread/performance.h	2026-10-05 11:31:06.686314840 +0200
+++ b/src/thread/performance.h	2026-10-05 12:09:37.104679724 +0200
@@ -1,12 +1,15 @@
 #pragma once 
 
 // apple doesnt provide jthread!
-#ifndef __APPLE__
+#if 1
 #include <thread>
 
 // NOTE only available on unix
 #include <assert.h>
 #include <sys/resource.h>
+#ifndef RUSAGE_THREAD
+#define RUSAGE_THREAD RUSAGE_SELF
+#endif
 #include <sys/socket.h>
 #include <sys/types.h>
 #include <sys/un.h>
